mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,804 @@
+
+/// Converts all imperfectly nested loop nests in `funcOp` to perfectly 
+/// nested loop nests by loop splitting.
+void LoopInterchange::handleImperfectlyNestedAffineLoops(Operation &funcOp) {
+  SmallVector<AffineForOp, 4> loopNest;
+  DenseMap<Operation *, SmallVector<AffineForOp, 4>> forTree;
+  DenseMap<Operation *, AffineForOp> forOperations;
+
+  // Stop splitting when each parent has only one child left.
+  bool oneChild = false;
+  while (!oneChild) {
+    oneChild = true;
+    // Walk the function to create a tree of affine.for operations.
+    funcOp.walk([&](AffineForOp op) {
+      loopNest.push_back(op);
+      if (op.getParentOp()->getName().getStringRef() == "affine.for")
+        forTree[op.getOperation()->getParentOp()].push_back(op);
+
+      forOperations[op.getOperation()] = op;
+    });
+    // Separate one of the sibling at a time.
+    for (auto &loopNest : forTree) {
+      // This loop nest has no sibling problem. Check the next loop nest.
+      if (loopNest.second.size() < 2)
+        continue;
+      oneChild = false;
+      separateSiblingAffineForOps(forOperations[loopNest.first],
+                                  loopNest.second.back(), loopNest.second);
+      // We need to walk the function again since the structure of loop nests
+      // within the funcOp body has changed.
+      break;
+    }
+    loopNest.clear();
+    forTree.clear();
+    forOperations.clear();
+  }
+  return;
+}
+
+/// Scans the loop nest rooted at `rootForOp` and collects all affine.load and
+/// affine.store ops. Fills `loadAndStoreOps` with all such ops.
+static void getAllLoadStores(AffineForOp rootForOp,
+                             SmallVector<Operation *, 8> &loadAndStoreOps) {
+  rootForOp.getOperation()->walk([&](Operation *op) {
+    if (isa<AffineLoadOp>(op) || isa<AffineStoreOp>(op)) {
+      loadAndStoreOps.push_back(op);
+    }
+  });
+}
+
+/// Fills `elementsSize` with size of element types of respective memrefs
+/// accessed by respective ops in `loadAndStoreOps`. These will be used to 
+/// check if two accesses are within a cache_line_size/element_size distance
+/// apart for a useful locality.
+static void getElementsSize(SmallVector<Operation *, 8> &loadAndStoreOps,
+                            DenseMap<Operation *, unsigned> &elementsSize) {
+  MemRefType memrefType;
+  for (Operation *op : loadAndStoreOps) {
+    if (isa<AffineLoadOp>(op)) {
+      AffineLoadOp loadOp = cast<AffineLoadOp>(*op);
+      memrefType = loadOp.getMemRefType();
+    } else if (isa<AffineStoreOp>(op)) {
+      AffineStoreOp storeOp = cast<AffineStoreOp>(*op);
+      memrefType = storeOp.getMemRefType();
+    }
+    // If the memref has a static shape, obtain the element size. Otherwise
+    // consider a default value.
+    elementsSize[op] = memrefType.hasStaticShape()
+                           ? getMemRefSizeInBytes(memrefType).getValue() /
+                                 memrefType.getNumElements()
+                           : 8;
```

Comment:
Don't hardcode this way. Define a static constant symbol in the pass and document it.

Reply:
Fixed. Thanks.